design/Fancy calling conventions-design.md

```diff
@@ -0,0 +1,118 @@
+# CSE231 Final Project Design
+
+Members:
+
+- Lien-Bee Huang (lih004@ucsd.edu)
+- Xi-Zhen Liu (xil087@ucsd.edu)
+- Yang-Hang Wu (yaw038@ucsd.edu)
+
+## Interfaces
+
+### AST
+
+```typescript=
+
+export type Parameter<A> = { 
+    name: string, 
+    type: Type,
+    value: Expr<A>,
+}
+export type FunDef<A> = { 
+    a?: A, 
+    name: string, 
+    parameters: Array<Parameter<A>>, 
+    ret: Type, 
+    inits: Array<VarInit<A>>, 
+    body: Array<Stmt<A>>,
+    arbarg_idx: number,
+    kwarg_idx: number,
+}
+```
+
+To deal with parameters with default value, we would like to add an attribute call `value` in `Parameter<A>` to store the default value. It could be any expression at the parsing stage. If there's no specific default value, the attribute will be set to `null`. This can also help us distinguish `None` from no default value.
+To deal with arbitrary arguments and keyword arguments, we would like to add attribute `arbarg_idx` to store the index of the tuple for the arbitrary arguments, and `kwarg_idx` to store the index of the dictionary for the keyword arguments. Since we will see all the arbitrary arguments as a tuple, and all the keyword arguments as a dictionary, we store the indices of them to distinguish them from normal tuple and dict parameters.
+
+### IR
+
+```typescript=
+import {Type, BinOp, UniOp} from './ast';
+
+export type Parameter<A> = { 
+    name: string, 
+    type: Type, 
+    value: Value<A> 
+}
+
+export type FunDef<A> = { 
+    a?: A, 
+    name: string, 
+    parameters: Array<Parameter<A>>, 
+    ret: Type, 
+    inits: Array<VarInit<A>>, 
+    body: Array<BasicBlock<A>>,
+    arbarg_idx: number,
+    kwarg_idx: number,
+}
+```
+
+The changes in IR are similar to AST, except that we could get the default value of parameters to be `Value<A>` type.
+
+## Testcases
+
+### Stage 1
+
+Support arguments with basic types and accept default value setting
+
+```python=
+def f(x : int, x : int = 5, z : int = 10): # redefine argument name
```

Comment:
Where are the body of these functions?

Reply:
I think this step (named args with names in assignment position) is probably enough for phase 1, **or** doing `*args` is enough for phase 1. I wouldn't try to do two of those first.

Reply:
We updated the testcases in this week progression

And our plan would be like the following:
- Stage 1 (This week)
  - default value in function definition
  - assign arguments by keywords
- Stage 2 (Future)
  - arbitrary arguments (it is dependent on tuple and dictionary, so we have to deal with this later)